### PR TITLE
Operations CLI framework

### PR DESCRIPTION
**Pull request by xbrianh**
This PR
  1. Provides a central entrypoint for DSS operations scripting, and
  2. Provides a framework for command forwarding to Lambda

connected to #2096, #2111

### CODE REVIEW

File: dss/operations/__init__.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            dispatch.actions[obj] = dict(target=dispatch.targets[self.target_name], name=name)
+            return obj
+        return register_action
+
+class DSSOperationsCommandDispatch:
+    targets: dict = dict()
+    actions: dict = dict()
+
+    def __init__(self):
+        self.parser = argparse.ArgumentParser()
```
Comment (DailyDreaming) [inline dss/operations/__init__.py]: Would it be a good idea to have a description here (example: https://github.com/HumanCellAtlas/data-store/blob/master/scripts/update_collection_db.py#L144 ) ?

**Comment by DailyDreaming** [comment]
Working now!  But the help menu is bit barren:
```
(v3nv) quokka@qcore:~/delete/data-store$ scripts/dss-ops.py --help
usage: dss-ops.py -h {} ...

positional arguments:
  {}

optional arguments:
  -h, --help  show this help message and exit
```

File: dss/operations/__init__.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                for argname in mutually_exclusive:
+                    kwargs = action_arguments.get(argname) or dict()
+                    group.add_argument(argname, **kwargs)
+            parser.add_argument("--job-id", default=dispatch.job_id)
+            parser.set_defaults(func=obj)
+            dispatch.actions[obj] = dict(target=dispatch.targets[self.target_name], name=name)
+            return obj
+        return register_action
+
+class DSSOperationsCommandDispatch:
```
Comment (amarjandu) [inline dss/operations/__init__.py]: if we place open bucket access to the lambda function, can we place a restriction here to authorized admins? we can use list: ADMIN_USER_EMAILS that we have in the lambda env var... not sure if this is available through our current invocation method for the lambda.

File: dss/operations/__init__.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+executed "in the cloud", via Lambda-SQS integration.
+
+Commands are defined with a target-action model. While command arguments configured at the
+target level are moved into the action arguments.
+
+e.g
+```
+dispatch = DSSOperationsCommandDispatch()
+
+storage = dispatch.target("storage", arguments={"replica", dict(choices=["dev", "integration", "staging", "prod"])}
```
Comment (DailyDreaming) [inline dss/operations/__init__.py]: Needs a closing parenthesis on the end.

File: dss/operations/__init__.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            logger.debug("Job ID: %s", args.job_id)
+            action_handler = args.func(argv, args) if isinstance(args.func, type) else args.func
+            try:
+                action_handler(argv, args)
+            except Exception:
+                logger.error(traceback.format_exc())
+        except SystemExit:
+            pass
+        except AttributeError:
+            print(self.__doc__)
```
Comment (DailyDreaming) [inline dss/operations/__init__.py]: Why does this print on an AttributeError?

File: dss/operations/__init__.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+if not os.environ.get("DSS_VERSION"):  # detect non-lambda environment
+    logger.setLevel(logging.DEBUG)
+    ch = logging.StreamHandler()
+    ch.setLevel(logging.DEBUG)
+    logger.addHandler(ch)
+
+class _target:
+    def __init__(self, target_name, dispatcher):
+        self.target_name = target_name
+        self.dispatch = dispatch
```
Comment (DailyDreaming) [inline dss/operations/__init__.py]: Should this be `self.dispatch = dispatcher`?

File: dss/operations/__init__.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    Central dispatch for the DSS Operations CLI.
+    """
+
+    targets: dict = dict()
+    actions: dict = dict()
+
+    def __init__(self):
+        self.parser = argparse.ArgumentParser(description=self.__doc__)
+        self.parser_targets = self.parser.add_subparsers()
+        self.job_id = str(uuid4())
```
Comment (DailyDreaming) [inline dss/operations/__init__.py]: How are job ids used?

File: dss/operations/__init__.py (hunk 5 above)
Comment (xbrianh) [reply dss/operations/__init__.py]: fixed

File: dss/operations/__init__.py (hunk 6 above)
Comment (xbrianh) [reply dss/operations/__init__.py]: It's used as a correlation id in log messages. Makes it easy to find results in cloudwatch

File: dss/operations/__init__.py (hunk 2 above)
Comment (xbrianh) [reply dss/operations/__init__.py]: An HCA AWS account will be required to communicate with the operations Lambda.

File: dss/operations/__init__.py (hunk 3 above)
Comment (xbrianh) [reply dss/operations/__init__.py]: fixed

File: dss/operations/__init__.py (hunk 4 above)
Comment (xbrianh) [reply dss/operations/__init__.py]: This makes it output help when no targets or actions are configured.

File: dss/operations/__init__.py (hunk 5 above)
Comment (xbrianh) [reply dss/operations/__init__.py]: fixed

File: dss/operations/__init__.py (hunk 1 above)
Comment (xbrianh) [reply dss/operations/__init__.py]: fixed

File: dss/operations/__init__.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+executed "in the cloud", via Lambda-SQS integration.
+
+Commands are defined with a target-action model. While command arguments configured at the
+target level are moved into the action arguments.
+
+e.g
+```
+dispatch = DSSOperationsCommandDispatch()
+
+storage = dispatch.target("storage", arguments={"replica": dict(choices=["dev", "integration", "staging", "prod"])})
```
Comment (DailyDreaming) [inline dss/operations/__init__.py]: `replica` should be `--replica` for the example execution command I think.

File: dss/operations/__init__.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+# execution:
+scripts/dss-ops.py storage verify-referential-integrity --replica staging
+```
+"""
+import os
+import argparse
+import logging
+import traceback
+from functools import partial
```
Comment (DailyDreaming) [inline dss/operations/__init__.py]: Unused import.

File: dss/operations/__init__.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+from uuid import uuid4
+
+import dss
+from dss import Replica
```
Comment (DailyDreaming) [inline dss/operations/__init__.py]: These appear unused:
`import dss`
`from dss import Replica`

Comment (xbrianh) [reply dss/operations/__init__.py]: fixed

File: dss/operations/__init__.py (hunk 8 above)
Comment (xbrianh) [reply dss/operations/__init__.py]: fixed

File: dss/operations/__init__.py (hunk 7 above)
Comment (xbrianh) [reply dss/operations/__init__.py]: Correct, fixed!

**Pull request merged**